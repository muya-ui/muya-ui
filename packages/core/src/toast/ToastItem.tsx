import React, { ReactNode } from 'react';
import styled, { css } from 'styled-components';

import { useEffectOnce } from '@muya-ui/utils';
import {
  ClearIcon,
  InformIcon,
  ReloadIcon,
  ReminderIcon,
  SuccessIcon,
} from '@muya-ui/theme-light';

import Spin from '../Spin';
import { IThemedBaseProps } from '../types';
import memoForwardRef from '../utils/memoForwardRef';
import useStyles from '../utils/useStyles';
import useTheme from '../utils/useTheme';
import { withThemeForStyled } from '../utils/withTheme';
import { IToastItemProps } from './types';

const StyledToastItem = styled.div`
  ${(props: IThemedBaseProps) => {
    const { theme } = props;
    const { colors, size } = theme;
    const { itemMarginBottom, shadow } = theme.components.Notification.toast;
    return css`
      display: flex;
      background: ${colors.pattern.background.higher};
      box-shadow: ${shadow};
      border-radius: ${size.spec.borderRadius.s2};
      pointer-events: all;
      margin-bottom: ${itemMarginBottom}px;
    `;
  }}
`;

const StyledContent = styled.div`
  ${(props: IThemedBaseProps) => {
    const { theme } = props;
    const { colors } = theme;
    const { contentStyle } = theme.components.Notification.toast;
    return css`
      box-sizing: border-box;
      color: ${colors.pattern.text.text};
      padding: ${contentStyle.padding};
      max-width: ${contentStyle.maxWidth}px;
      min-width: ${contentStyle.minWidth}px;
      line-height: ${contentStyle.lineHeight}px;
      font-size: ${contentStyle.fontSize}px;
    `;
  }}
`;

const StyledIcon = styled.div`
  box-sizing: border-box;
  ${(props: IThemedBaseProps) => {
    const { theme } = props;
    const { iconStyle } = theme.components.Notification.toast;
    return css`
      padding: ${iconStyle.padding};
      font-size: ${iconStyle.size}px;
      line-height: ${iconStyle.size}px;
      width: ${iconStyle.width}px;
    `;
  }}
`;

const BaseNode = styled.div``;
const ToastItemWithoutStyle = memoForwardRef<HTMLDivElement, IToastItemProps>((props, ref) => {
  const { item, timeout, state, className, style = {}, ...otherProps } = props;
  const theme = useTheme();
  const innerStyles = useStyles(
    'toast-item',
    {
      self: '',
      container: '',
      icon: '',
      content: '',
    },
    item.styles,
  );
  let iconNode: ReactNode;
  if (item.icon) {
    iconNode = item.icon;
  } else if (item.type !== 'loading') {
    const { Notification: token } = theme.components;
    const tokenIcon = {
      info: InformIcon,
      error: ClearIcon,
      success: SuccessIcon,
      warning: ReminderIcon,
      loading: ReloadIcon,
      ...token.icon,
    };
    const Icon = tokenIcon[item.type];
    const bgColor = token.iconBgColor[item.type];
    iconNode = (
      <StyledIcon {...innerStyles.icon} theme={theme}>
        <Icon bgColor={bgColor} />
      </StyledIcon>
    );
  } else {
    iconNode = (
      <StyledIcon {...innerStyles.icon} theme={theme}>
        <Spin />
      </StyledIcon>
    );
  }
  let contentNode: ReactNode;
  if (typeof item.content === 'string' || item.enableContentWrapper) {
    contentNode = (
      <StyledContent {...innerStyles.content} theme={theme}>
        {item.content}
      </StyledContent>
    );
  } else {
    contentNode = item.content;
  }
  const containerProps = {
    style: {
      ...innerStyles.self.style,
      ...style,
    },
    className: [className, innerStyles.self.className].join(' '),
    ...otherProps,
  };

  useEffectOnce(() => () => {
    if (item.onClose) {
      item.onClose();
    }
  });
  return (
    <BaseNode ref={ref} {...containerProps}>
      <StyledToastItem theme={theme} {...innerStyles.container}>
        {iconNode}
        {contentNode}
      </StyledToastItem>
    </BaseNode>
  );
});

const ToastItem = styled(ToastItemWithoutStyle)<IToastItemProps>`
  ${props => {
    const { theme, item, state, timeout } = props;
    const { colors, transition } = theme;
    const { lineMaxChar, contentStyle, minHeight } = theme.components.Notification.toast;
    let maxHeight = item.maxHeight || minHeight;
    if (typeof item.content === 'string' && item.content.length > lineMaxChar) {
      const lineNum = Math.ceil(item.content.length / lineMaxChar);
      maxHeight = minHeight + lineNum * contentStyle.lineHeight;
    }
    const transitionMap = {
      entering: '',
      entered: css`
        max-height: ${maxHeight}px;
        opacity: 1;
      `,
      exiting: css`
        max-height: 0;
        opacity: 0;
        transition-timing-function: ${transition.pattern.easing.leave};
      `,
      exited: '',
      unmounted: '',
    };
    const transformCSS =
      transitionMap[state] ||
      css`
        max-height: 0;
        opacity: 0;
      `;
    return css`
      transition: all ${timeout}ms ${transition.pattern.easing.enter};
      pointer-events: none;
      display: flex;
      align-items: center;
      justify-content: center;
      ${transformCSS}

      & ${StyledIcon} {
        color: ${colors.pattern.feature[item.type!]};
      }
    `;
  }}
`;

export default withThemeForStyled(ToastItem);
