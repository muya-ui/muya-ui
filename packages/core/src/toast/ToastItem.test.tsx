import React from 'react';
import renderer from 'react-test-renderer';

import ToastItem from './ToastItem';

test('测试普通渲染', () => {
  const tree = renderer
    .create(
      <>
        <ToastItem timeout={100} state="entering" item={{ type: 'success', content: 'ssss' }} />
        <ToastItem timeout={100} state="entered" item={{ type: 'loading', content: 'ssss' }} />
        <ToastItem timeout={100} state="exited" item={{ type: 'error', content: 'ssss' }} />
        <ToastItem timeout={100} state="exiting" item={{ type: 'error', content: 'ssss' }} />
      </>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('测试 自定义节点', () => {
  const tree = renderer
    .create(
      <ToastItem
        state="entering"
        timeout={100}
        item={{
          type: 'success',
          content: <div>content</div>,
          icon: <div>icon</div>,
        }}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
