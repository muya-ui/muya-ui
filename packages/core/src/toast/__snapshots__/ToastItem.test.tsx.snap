// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 自定义节点 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgba(255,255,255,1);
  box-shadow: 0 0 24px 0 rgba(102,102,102,0.08);
  border-radius: 4px;
  pointer-events: all;
  margin-bottom: 16px;
}

.c0 {
  -webkit-transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  pointer-events: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-height: 0;
  opacity: 0;
}

.c0 .c2 {
  color: rgba(0,217,72,1);
}

<div
  className="c0 muya-toast-item-self"
  style={Object {}}
>
  <div
    className="c1 muya-toast-item-container"
  >
    <div>
      icon
    </div>
    <div>
      content
    </div>
  </div>
</div>
`;

exports[`测试普通渲染 1`] = `
Array [
  .c4 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgba(255,255,255,1);
  box-shadow: 0 0 24px 0 rgba(102,102,102,0.08);
  border-radius: 4px;
  pointer-events: all;
  margin-bottom: 16px;
}

.c5 {
  box-sizing: border-box;
  color: rgba(102,102,102,1);
  padding: 12px 12px 12px 0;
  max-width: 488px;
  min-width: 294px;
  line-height: 16px;
  font-size: 12px;
}

.c3 {
  box-sizing: border-box;
  padding: 11px 8px 0 12px;
  font-size: 16px;
  line-height: 16px;
  width: 36px;
}

.c0 {
  -webkit-transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  pointer-events: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-height: 0;
  opacity: 0;
}

.c0 .c2 {
  color: rgba(0,217,72,1);
}

.c6 .c2 {
  color: rgba(26,122,248,1);
}

.c7 .c2 {
  color: rgba(255,43,0,1);
}

.c8 .c2 {
  color: rgba(255,43,0,1);
}

<div
    className="c0 muya-toast-item-self"
    style={Object {}}
  >
    <div
      className="c1 muya-toast-item-container"
    >
      <div
        className="c2 c3 muya-toast-item-icon"
      >
        <svg
          aria-hidden="false"
          className="c4"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M436.576 574.933l-105.6-105.6a21.333 21.333 0 00-30.165 0l-60.342 60.342a21.333 21.333 0 000 30.165l181.014 181.013c4.17 4.171 9.632 6.262 15.093 6.262s10.923-2.091 15.083-6.251L783.53 408.992a21.333 21.333 0 000-30.165l-60.342-60.342a21.333 21.333 0 00-30.165 0L436.576 574.923zM512 1024C229.227 1024 0 794.773 0 512S229.227 0 512 0s512 229.227 512 512-229.227 512-512 512z"
          />
        </svg>
      </div>
      <div
        className="c5 muya-toast-item-content"
      >
        ssss
      </div>
    </div>
  </div>,
  .c5 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c6 {
  -webkit-animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
  animation: iECmZH 0.75s linear infinite,keNNsN 1.5s linear infinite;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgba(255,255,255,1);
  box-shadow: 0 0 24px 0 rgba(102,102,102,0.08);
  border-radius: 4px;
  pointer-events: all;
  margin-bottom: 16px;
}

.c7 {
  box-sizing: border-box;
  color: rgba(102,102,102,1);
  padding: 12px 12px 12px 0;
  max-width: 488px;
  min-width: 294px;
  line-height: 16px;
  font-size: 12px;
}

.c3 {
  box-sizing: border-box;
  padding: 11px 8px 0 12px;
  font-size: 16px;
  line-height: 16px;
  width: 36px;
}

.c8 .c2 {
  color: rgba(0,217,72,1);
}

.c0 {
  -webkit-transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  pointer-events: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-height: 60px;
  opacity: 1;
}

.c0 .c2 {
  color: rgba(26,122,248,1);
}

.c9 .c2 {
  color: rgba(255,43,0,1);
}

.c10 .c2 {
  color: rgba(255,43,0,1);
}

<div
    className="c0 muya-toast-item-self"
    style={Object {}}
  >
    <div
      className="c1 muya-toast-item-container"
    >
      <div
        className="c2 c3 muya-toast-item-icon"
      >
        <span
          className="c4"
          direction="column"
        >
          <span
            className=""
          >
            <svg
              aria-hidden="false"
              className="c5 c6"
              focusable="false"
              preserveAspectRatio="xMidYMid"
              viewBox="0 0 100 100"
            >
              <circle
                cx="50"
                cy="50"
                fill="none"
                r="45"
                stroke="rgba(26,122,248,1)"
                strokeLinecap="round"
                strokeWidth="10"
                transform="rotate(144 50 50)"
              />
            </svg>
          </span>
          <span />
        </span>
      </div>
      <div
        className="c7 muya-toast-item-content"
      >
        ssss
      </div>
    </div>
  </div>,
  .c4 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgba(255,255,255,1);
  box-shadow: 0 0 24px 0 rgba(102,102,102,0.08);
  border-radius: 4px;
  pointer-events: all;
  margin-bottom: 16px;
}

.c5 {
  box-sizing: border-box;
  color: rgba(102,102,102,1);
  padding: 12px 12px 12px 0;
  max-width: 488px;
  min-width: 294px;
  line-height: 16px;
  font-size: 12px;
}

.c3 {
  box-sizing: border-box;
  padding: 11px 8px 0 12px;
  font-size: 16px;
  line-height: 16px;
  width: 36px;
}

.c6 .c2 {
  color: rgba(0,217,72,1);
}

.c7 .c2 {
  color: rgba(26,122,248,1);
}

.c0 {
  -webkit-transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  pointer-events: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-height: 0;
  opacity: 0;
}

.c0 .c2 {
  color: rgba(255,43,0,1);
}

.c8 .c2 {
  color: rgba(255,43,0,1);
}

<div
    className="c0 muya-toast-item-self"
    style={Object {}}
  >
    <div
      className="c1 muya-toast-item-container"
    >
      <div
        className="c2 c3 muya-toast-item-icon"
      >
        <svg
          aria-hidden="false"
          className="c4"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M602.507 512L730.73 383.787a21.333 21.333 0 000-30.187l-60.342-60.33a21.333 21.333 0 00-30.165 0L512 421.482 383.787 293.269a21.333 21.333 0 00-30.187 0l-60.33 60.342a21.333 21.333 0 000 30.165L421.482 512 293.269 640.213a21.333 21.333 0 000 30.187l60.342 60.33a21.333 21.333 0 0030.165 0L512 602.518l128.213 128.214a21.333 21.333 0 0030.187 0l60.33-60.342a21.333 21.333 0 000-30.165L602.518 512zM512 1024C229.227 1024 0 794.773 0 512S229.227 0 512 0s512 229.227 512 512-229.227 512-512 512z"
          />
        </svg>
      </div>
      <div
        className="c5 muya-toast-item-content"
      >
        ssss
      </div>
    </div>
  </div>,
  .c4 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  fill: currentColor;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgba(255,255,255,1);
  box-shadow: 0 0 24px 0 rgba(102,102,102,0.08);
  border-radius: 4px;
  pointer-events: all;
  margin-bottom: 16px;
}

.c5 {
  box-sizing: border-box;
  color: rgba(102,102,102,1);
  padding: 12px 12px 12px 0;
  max-width: 488px;
  min-width: 294px;
  line-height: 16px;
  font-size: 12px;
}

.c3 {
  box-sizing: border-box;
  padding: 11px 8px 0 12px;
  font-size: 16px;
  line-height: 16px;
  width: 36px;
}

.c6 .c2 {
  color: rgba(0,217,72,1);
}

.c7 .c2 {
  color: rgba(26,122,248,1);
}

.c8 .c2 {
  color: rgba(255,43,0,1);
}

.c0 {
  -webkit-transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  transition: all 100ms cubic-bezier(0.0,0,0.2,1);
  pointer-events: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-height: 0;
  opacity: 0;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,1,1);
  transition-timing-function: cubic-bezier(0.4,0,1,1);
}

.c0 .c2 {
  color: rgba(255,43,0,1);
}

<div
    className="c0 muya-toast-item-self"
    style={Object {}}
  >
    <div
      className="c1 muya-toast-item-container"
    >
      <div
        className="c2 c3 muya-toast-item-icon"
      >
        <svg
          aria-hidden="false"
          className="c4"
          focusable="false"
          viewBox="0 0 1024 1024"
        >
          <rect
            fill="rgba(255,255,255,1)"
            height={716.8}
            width={716.8}
            x={153.60000000000002}
            y={153.60000000000002}
          />
          <path
            d="M602.507 512L730.73 383.787a21.333 21.333 0 000-30.187l-60.342-60.33a21.333 21.333 0 00-30.165 0L512 421.482 383.787 293.269a21.333 21.333 0 00-30.187 0l-60.33 60.342a21.333 21.333 0 000 30.165L421.482 512 293.269 640.213a21.333 21.333 0 000 30.187l60.342 60.33a21.333 21.333 0 0030.165 0L512 602.518l128.213 128.214a21.333 21.333 0 0030.187 0l60.33-60.342a21.333 21.333 0 000-30.165L602.518 512zM512 1024C229.227 1024 0 794.773 0 512S229.227 0 512 0s512 229.227 512 512-229.227 512-512 512z"
          />
        </svg>
      </div>
      <div
        className="c5 muya-toast-item-content"
      >
        ssss
      </div>
    </div>
  </div>,
]
`;
